File: src/main/java/com/docs/viewer/common/file/api/FileController.java
package com.docs.viewer.common.file.api;

import com.docs.viewer.common.file.dto.FileAttachInfo;
import com.docs.viewer.common.file.dto.request.FileUploadRequest;
import com.docs.viewer.common.file.dto.response.FileResponse;
import com.docs.viewer.common.file.service.FileService;
import com.docs.viewer.common.preview.service.PreviewService;
import com.docs.viewer.global.common.dto.response.ResponseHandler;
import com.docs.viewer.global.common.utils.FileTypeUtil;
import com.docs.viewer.global.error.dto.ErrorCode;
import com.docs.viewer.global.error.dto.ErrorInfo;
import com.docs.viewer.global.error.dto.response.ErrorHandler;
import com.docs.viewer.global.error.dto.response.ErrorResponse;
import lombok.RequiredArgsConstructor;
import org.apache.tika.Tika;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import jakarta.validation.Valid;

import java.net.URI;
import java.util.List;

@RestController
@RequiredArgsConstructor
@RequestMapping(value = "/file")
public class FileController {

    private final FileService fileService;
    private final PreviewService previewService;
    private final ErrorHandler errorHandler;

    @GetMapping
    public ResponseEntity<?> fileImage(@RequestParam Integer attachFile) {
        try {
            FileResponse fileResponse = this.findFile(attachFile);
            String mimeType = fileResponse.getMimeType();
            if (FileTypeUtil.isAllowType(mimeType)) {
                ByteArrayResource resource = previewService.findFile(fileResponse);
                MediaType mediaType = MediaType.parseMediaType(mimeType);
                return ResponseEntity.ok().contentType(mediaType)
                        .body(resource);
            }
            return badRequest(errorHandler.buildError(ErrorCode.FILETYPE_MAPPING_INVALID, ErrorInfo.builder()
                    .errors(mimeType)
                    .build()));
        } catch (Exception e) {
            return unprocessableEntity(errorHandler.buildError(ErrorCode.INTERNAL_SERVER_ERROR, ErrorInfo.builder()
                    .message(e.getMessage())
                    .build()));
        }
    }

    @PostMapping(value = "/upload", consumes = {MediaType.MULTIPART_FORM_DATA_VALUE})
    public ResponseEntity<?> uploadFiles(@ModelAttribute @Valid FileUploadRequest request) {
        try {
            List<MultipartFile> files = request.getFiles();
            List<FileAttachInfo> fileAttachInfos = fileService.saveFile(files);
            for (FileAttachInfo fileAttachInfo : fileAttachInfos) {
                Integer attachFile = fileAttachInfo.getFileAttachCode();
                FileResponse fileResponse = this.findFile(attachFile);
                previewService.saveFile(attachFile, fileResponse);
            }
            return ResponseEntity.created(URI.create("/create"))
                    .body(ResponseHandler.builder()
                            .status(HttpStatus.CREATED.value())
                            .message("파일이 업로드 되었습니다")
                            .build());
        } catch (Exception e) {
            return unprocessableEntity(errorHandler.buildError(ErrorCode.INTERNAL_SERVER_ERROR, ErrorInfo.builder()
                    .message(e.getMessage())
                    .build()));
        }
    }

    private FileResponse findFile(Integer attachFile) throws Exception {
        byte[] fileBytes = fileService.findFile(attachFile);
        ByteArrayResource resource = new ByteArrayResource(fileBytes);
        String mimeType = this.findMimeType(fileBytes);
        return FileResponse.of(resource, mimeType);
    }

    private String findMimeType(byte[] fileBytes) {
        Tika tika = new Tika();
        return tika.detect(fileBytes);
    }

    private ResponseEntity<ResponseHandler> badRequest(ErrorResponse response) {
        return ResponseEntity.badRequest().body(ResponseHandler.error(response));
    }

    private ResponseEntity<ErrorResponse> unprocessableEntity(ErrorResponse response) {
        return ResponseEntity.unprocessableEntity().body(response);
    }
}

File: dotnet-docs-viewer/Controllers/FileController.cs
using System.Net.Mime;
using Common.Utils;
using HeyRed.Mime;
using Microsoft.AspNetCore.Mvc;
using WebApp.Common.Dto;
using WebApp.Common.Error;
using WebApp.Common.Error.Types;
using WebApp.Dto;
using WebApp.Service;

namespace WebApp.Controllers
{
    [ApiController]
    [Route("file")]
    public class FileController : ControllerBase
    {
        private readonly IFileService _fileService;
        private readonly IPreviewService _previewService;
        private readonly IErrorHandler _errorHandler;

        public FileController(IFileService fileService, IPreviewService previewService, IErrorHandler errorHandler)
        {
            _fileService = fileService;
            _previewService = previewService;
            _errorHandler = errorHandler;
        }

        [HttpGet]
        public IActionResult FileImage([FromQuery] int attachFile)
        {
            try
            {
                var fileResponse = FindFileAsync(attachFile).Result;;
                var mimeType = fileResponse.MimeType;

                if (FileTypeUtil.IsAllowType(mimeType))
                {
                    byte[] resource = _previewService.FindFileAsync(fileResponse).Result;

                    return File(resource, mimeType);
                }

                return BadRequest(ResponseHandler.Error(_errorHandler.BuildError(
                    ErrorCode.FILETYPE_MAPPING_INVALID,
                    new ErrorInfo { Errors = mimeType })
                ));
            }
            catch (Exception ex)
            {
                return UnprocessableEntity(_errorHandler.BuildError(
                    ErrorCode.INTERNAL_SERVER_ERROR,
                    new ErrorInfo { Message = ex.Message }
                ));
            }
        }

        [HttpPost("upload")]
        [Consumes("multipart/form-data")]
        public async Task<IActionResult> UploadFiles([FromForm] FileUploadRequest request)
        {
            try
            {
                var files = request.Files;
                var fileAttachInfos = await _fileService.SaveFileAsync(files);

                foreach (var fileAttachInfo in fileAttachInfos)
                {
                    var attachFile = fileAttachInfo.FileAttachCode;
                    var fileResponse = FindFileAsync(attachFile).Result;

                    await _previewService.SaveFileAsync(attachFile, fileResponse);
                }

                return Created("/create", ResponseHandler.Success(
                    StatusCodes.Status201Created, 
                    "파일이 업로드 되었습니다"
                ));
            }
            catch (Exception ex)
            {
                var error = _errorHandler.BuildError(
                    ErrorCode.INTERNAL_SERVER_ERROR,
                    new ErrorInfo { Message = ex.Message });

                return UnprocessableEntity(error);
            }
        }

        private async Task<FileResponse> FindFileAsync(int attachFile)
        {
            byte[] fileBytes = await _fileService.FindFileAsync(attachFile);
            string mimeType = FindMimeType(fileBytes);

            return FileResponse.Of(fileBytes, mimeType);
        }

        private string FindMimeType(byte[] fileBytes)
        {
            try
            {
                return MimeGuesser.GuessMimeType(fileBytes);
            }
            catch
            {
                return MediaTypeNames.Application.Octet;
            }
        }
    }
}